registry/storage/driver/swift/swift.go

```diff
@@ -0,0 +1,561 @@
+	}
+
+	if d.BulkDeleteSupport {
+		if _, err := d.Conn.BulkDelete(d.Container, objects); err != swift.Forbidden {
+			return parseError(path, err)
+		}
+	}
+
+	for _, name := range objects {
+		if _, headers, err := d.Conn.Object(d.Container, name); err == nil {
+			manifest, ok := headers["X-Object-Manifest"]
+			if ok {
+				components := strings.SplitN(manifest, "/", 2)
+				segContainer := components[0]
+				segments, err := d.getAllSegments(components[1])
+				if err != nil {
+					return parseError(name, err)
+				}
+
+				for _, s := range segments {
+					if err := d.Conn.ObjectDelete(segContainer, s.Name); err != nil {
+						return parseError(s.Name, err)
+					}
+				}
+			}
+		} else {
+			return parseError(name, err)
+		}
+
+		if err := d.Conn.ObjectDelete(d.Container, name); err != nil {
+			return parseError(name, err)
+		}
+	}
+
+	return nil
+}
+
+// URLFor returns a URL which may be used to retrieve the content stored at the given path.
+// May return an UnsupportedMethodErr in certain StorageDriver implementations.
+func (d *driver) URLFor(ctx context.Context, path string, options map[string]interface{}) (string, error) {
+	return "", storagedriver.ErrUnsupportedMethod
+}
+
+func (d *driver) swiftPath(path string) string {
+	return strings.TrimLeft(strings.TrimRight(d.Prefix+"/files"+path, "/"), "/")
+}
+
+func (d *driver) swiftSegmentPath(path string) string {
+	return strings.TrimLeft(strings.TrimRight(d.Prefix+"/segments"+path, "/"), "/")
+}
+
+func (d *driver) createParentFolders(path string) error {
+	dir := gopath.Dir(path)
+	for dir != "/" {
+		_, _, err := d.Conn.Object(d.Container, d.swiftPath(dir))
+		if err == swift.ContainerNotFound || err == swift.ObjectNotFound {
+			_, err := d.Conn.ObjectPut(d.Container, d.swiftPath(dir), bytes.NewReader(make([]byte, 0)),
+				false, "", directoryMimeType, nil)
+			if err != nil {
+				return parseError(dir, err)
+			}
+		}
+		dir = gopath.Dir(dir)
+	}
+
+	return nil
+}
+
+func (d *driver) getContentType() string {
+	return "application/octet-stream"
+}
+
+func (d *driver) getAllSegments(path string) ([]swift.Object, error) {
+	return d.Conn.Objects(d.Container, &swift.ObjectsOpts{Prefix: d.swiftSegmentPath(path)})
+}
+
+func (d *driver) createManifest(path string) (*swift.ObjectCreateFile, error) {
+	headers := make(swift.Headers)
+	headers["X-Object-Manifest"] = d.Container + "/" + d.swiftSegmentPath(path)
+	return d.Conn.ObjectCreate(d.Container, d.swiftPath(path), false, "",
+		d.getContentType(), headers)
+}
+
+func detectBulkDelete(authURL string) (bulkDelete bool) {
+	resp, err := http.Get(filepath.Join(authURL, "..", "..") + "/info")
+	if err == nil {
+		defer resp.Body.Close()
+		decoder := json.NewDecoder(resp.Body)
+		var infos swiftInfo
+		if decoder.Decode(&infos) == nil {
+			_, bulkDelete = infos["bulk_delete"]
+		}
+	}
+	return
+}
+
+func parseError(path string, err error) error {
+	if err == swift.ContainerNotFound || err == swift.ObjectNotFound {
```

Comment:
It's true that the deletion of the container is a more severe error that just a parent folder missing, I think it will still fail "properly", with the PathNotFound error though. We could fail with an "storagedriver.InvalidPathError" but the documentation says it is used for malformed path. What do you think ?